standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java

```diff
@@ -281,9 +280,14 @@ public void markCompacted(CompactionInfo info) throws MetaException {
       try {
         dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
         stmt = dbConn.createStatement();
-        String s = "SELECT \"CQ_ID\", \"CQ_DATABASE\", \"CQ_TABLE\", \"CQ_PARTITION\", " +
-            "\"CQ_TYPE\", \"CQ_RUN_AS\", \"CQ_HIGHEST_WRITE_ID\" FROM \"COMPACTION_QUEUE\" " +
-            "WHERE \"CQ_STATE\" = '" + READY_FOR_CLEANING + "'";
+        /*
+         * By filtering on minOpenTxnWaterMark, we will only cleanup after every transaction is committed, that could see
+         * the uncompacted deltas. This way the cleaner can clean up everything that was made obsolete by this compaction.
+         */
+        long minOpenTxnWaterMark = getMinOpenTxnIdWaterMark(dbConn);
```

Comment:
Cleaner already knows this value, Cleaner#run calls CompactionTxnHandler#findMinOpenTxnIdForCleaner first, then findReadyToClean, so you can just pass it into findReadyToClean.

(Btw findMinOpenTxnIdForCleaner doesn't filter out timed out txns like getMinOpenTxnIdWaterMark does, might want to change that? (AcidHouseKeeperService should take care of that, but who knows if it's on... on the other hand that's another query and would take longer))

Reply:
1. Passing the minOpenTxn as an argument now
2. Changed the findMinOpenTxnIdForCleaner to use getMinOpenTxnIdWaterMark. The timeout boundary checking is needed, since HIVE-23084, because it might be possible for an open txn to appear later, that has txnId lower than the current minOpen and higher the timeout boundary. Probably it wouldn't cause any problem for the Cleaner, but better safe than sorry, this way it always gives correct result.
This also means that the max(cq_next_txnid) check is removed, but I think this will only mean, that if there were any txn after the compaction that were aborted, we are going to clean those up also, which is a good side effect.